### ISSUE TITLE
CI failing due to removed dataset IDs from OpenML

### ISSUE THREAD

**Issue by adrinjalali** [fairlearn/fairlearn#703]
The commit 77e51c5 seems to be triggering a failure in our CI (see https://github.com/fairlearn/fairlearn/pull/702) caused by one or two IDs remvoed from OpenML. We should investigate the issue and figure out why they're removed, and where they can be found now.

<details>

```
2021-02-19T05:31:46.6084363Z =================================== FAILURES ===================================
2021-02-19T05:31:46.6191931Z _________ TestFairlearnDataset.test_dataset_as_bunch[fetch_adult-True] _________
2021-02-19T05:31:46.6192578Z 
2021-02-19T05:31:46.6193185Z self = <test.unit.datasets.test_datasets.TestFairlearnDataset object at 0x7fdc98376040>
2021-02-19T05:31:46.6194599Z as_frame = True, fetch_function = <function fetch_adult at 0x7fdc9e145430>
2021-02-19T05:31:46.6195290Z 
2021-02-19T05:31:46.6196268Z     @pytest.mark.parametrize("as_frame", [True, False])
2021-02-19T05:31:46.6197361Z     @pytest.mark.parametrize("fetch_function", [fetch_adult, fetch_boston, fetch_bank_marketing])
2021-02-19T05:31:46.6198700Z     def test_dataset_as_bunch(self, as_frame, fetch_function):
2021-02-19T05:31:46.6199841Z >       dataset = fetch_function(as_frame=as_frame)
2021-02-19T05:31:46.6200488Z 
2021-02-19T05:31:46.6201319Z test/unit/datasets/test_datasets.py:18: 
2021-02-19T05:31:46.6203125Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-02-19T05:31:46.6204409Z fairlearn/datasets/_fetch_adult.py:79: in fetch_adult
2021-02-19T05:31:46.6205231Z     return fetch_openml(
2021-02-19T05:31:46.6206504Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/utils/validation.py:63: in inner_f
2021-02-19T05:31:46.6213353Z     return f(*args, **kwargs)
2021-02-19T05:31:46.6214531Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:847: in fetch_openml
2021-02-19T05:31:46.6216700Z     data_description = _get_data_description_by_id(data_id, data_home)
2021-02-19T05:31:46.6217842Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:437: in _get_data_description_by_id
2021-02-19T05:31:46.6219080Z     json_data = _get_json_content_from_openml_api(
2021-02-19T05:31:46.6220097Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-02-19T05:31:46.6220979Z 
2021-02-19T05:31:46.6221694Z url = 'api/v1/json/data/1590'
2021-02-19T05:31:46.6223169Z error_message = 'Dataset with data_id 1590 not found.'
2021-02-19T05:31:46.6223991Z data_home = '/home/vsts/.fairlearn-data/openml'
2021-02-19T05:31:46.6224318Z 
2021-02-19T05:31:46.6224799Z     def _get_json_content_from_openml_api(
2021-02-19T05:31:46.6225233Z         url: str,
2021-02-19T05:31:46.6225693Z         error_message: Optional[str],
2021-02-19T05:31:46.6226170Z         data_home: Optional[str]
2021-02-19T05:31:46.6226760Z     ) -> Dict:
2021-02-19T05:31:46.6227159Z         """
2021-02-19T05:31:46.6227588Z         Loads json data from the openml api
2021-02-19T05:31:46.6228012Z     
2021-02-19T05:31:46.6228377Z         Parameters
2021-02-19T05:31:46.6228996Z         ----------
2021-02-19T05:31:46.6229365Z         url : str
2021-02-19T05:31:46.6229793Z             The URL to load from. Should be an official OpenML endpoint
2021-02-19T05:31:46.6230186Z     
2021-02-19T05:31:46.6230549Z         error_message : str or None
2021-02-19T05:31:46.6231014Z             The error message to raise if an acceptable OpenML error is thrown
2021-02-19T05:31:46.6231726Z             (acceptable error is, e.g., data id not found. Other errors, like 404's
2021-02-19T05:31:46.6232227Z             will throw the native error message)
2021-02-19T05:31:46.6232585Z     
2021-02-19T05:31:46.6233132Z         data_home : str or None
2021-02-19T05:31:46.6233587Z             Location to cache the response. None if no cache is required.
2021-02-19T05:31:46.6233996Z     
2021-02-19T05:31:46.6234307Z         Returns
2021-02-19T05:31:46.6234784Z         -------
2021-02-19T05:31:46.6235163Z         json_data : json
2021-02-19T05:31:46.6235708Z             the json result from the OpenML server if the call was successful.
2021-02-19T05:31:46.6236153Z             An exception otherwise.
2021-02-19T05:31:46.6236520Z         """
2021-02-19T05:31:46.6236841Z     
2021-02-19T05:31:46.6237212Z         @_retry_with_clean_cache(url, data_home)
2021-02-19T05:31:46.6237609Z         def _load_json():
2021-02-19T05:31:46.6238083Z             with closing(_open_openml_url(url, data_home)) as response:
2021-02-19T05:31:46.6238783Z                 return json.loads(response.read().decode("utf-8"))
2021-02-19T05:31:46.6239198Z     
2021-02-19T05:31:46.6239512Z         try:
2021-02-19T05:31:46.6239875Z             return _load_json()
2021-02-19T05:31:46.6240283Z         except HTTPError as error:
2021-02-19T05:31:46.6240757Z             # 412 is an OpenML specific error code, indicating a generic error
2021-02-19T05:31:46.6241199Z             # (e.g., data not found)
2021-02-19T05:31:46.6241611Z             if error.code != 412:
2021-02-19T05:31:46.6242010Z                 raise error
2021-02-19T05:31:46.6242508Z     
2021-02-19T05:31:46.6242877Z         # 412 error, not in except for nicer traceback
2021-02-19T05:31:46.6243310Z >       raise OpenMLError(error_message)
2021-02-19T05:31:46.6243799Z E       sklearn.datasets._openml.OpenMLError: Dataset with data_id 1590 not found.
2021-02-19T05:31:46.6244122Z 
2021-02-19T05:31:46.6244826Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:179: OpenMLError
2021-02-19T05:31:46.6245596Z ____ TestFairlearnDataset.test_dataset_as_bunch[fetch_bank_marketing-True] _____
2021-02-19T05:31:46.6245912Z 
2021-02-19T05:31:46.6246337Z self = <test.unit.datasets.test_datasets.TestFairlearnDataset object at 0x7fdc98293e20>
2021-02-19T05:31:46.6246782Z as_frame = True
2021-02-19T05:31:46.6247203Z fetch_function = <function fetch_bank_marketing at 0x7fdc9e145160>
2021-02-19T05:31:46.6247483Z 
2021-02-19T05:31:46.6247859Z     @pytest.mark.parametrize("as_frame", [True, False])
2021-02-19T05:31:46.6248376Z     @pytest.mark.parametrize("fetch_function", [fetch_adult, fetch_boston, fetch_bank_marketing])
2021-02-19T05:31:46.6248928Z     def test_dataset_as_bunch(self, as_frame, fetch_function):
2021-02-19T05:31:46.6249411Z >       dataset = fetch_function(as_frame=as_frame)
2021-02-19T05:31:46.6249668Z 
2021-02-19T05:31:46.6250030Z test/unit/datasets/test_datasets.py:18: 
2021-02-19T05:31:46.6250489Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-02-19T05:31:46.6251024Z fairlearn/datasets/_fetch_bank_marketing.py:82: in fetch_bank_marketing
2021-02-19T05:31:46.6251455Z     return fetch_openml(
2021-02-19T05:31:46.6252171Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/utils/validation.py:63: in inner_f
2021-02-19T05:31:46.6252687Z     return f(*args, **kwargs)
2021-02-19T05:31:46.6253417Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:873: in fetch_openml
2021-02-19T05:31:46.6253972Z     features_list = _get_data_features(data_id, data_home)
2021-02-19T05:31:46.6254801Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:450: in _get_data_features
2021-02-19T05:31:46.6255366Z     json_data = _get_json_content_from_openml_api(
2021-02-19T05:31:46.6257168Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-02-19T05:31:46.6257396Z 
2021-02-19T05:31:46.6257820Z url = 'api/v1/json/data/features/1461'
2021-02-19T05:31:46.6258301Z error_message = 'Dataset with data_id 1461 not found.'
2021-02-19T05:31:46.6258783Z data_home = '/home/vsts/.fairlearn-data/openml'
2021-02-19T05:31:46.6259060Z 
2021-02-19T05:31:46.6259316Z     def _get_json_content_from_openml_api(
2021-02-19T05:31:46.6259600Z         url: str,
2021-02-19T05:31:46.6259873Z         error_message: Optional[str],
2021-02-19T05:31:46.6260181Z         data_home: Optional[str]
2021-02-19T05:31:46.6260577Z     ) -> Dict:
2021-02-19T05:31:46.6260885Z         """
2021-02-19T05:31:46.6261154Z         Loads json data from the openml api
2021-02-19T05:31:46.6261415Z     
2021-02-19T05:31:46.6261626Z         Parameters
2021-02-19T05:31:46.6261995Z         ----------
2021-02-19T05:31:46.6262254Z         url : str
2021-02-19T05:31:46.6262578Z             The URL to load from. Should be an official OpenML endpoint
2021-02-19T05:31:46.6262866Z     
2021-02-19T05:31:46.6263120Z         error_message : str or None
2021-02-19T05:31:46.6263482Z             The error message to raise if an acceptable OpenML error is thrown
2021-02-19T05:31:46.6264060Z             (acceptable error is, e.g., data id not found. Other errors, like 404's
2021-02-19T05:31:46.6264451Z             will throw the native error message)
2021-02-19T05:31:46.6264720Z     
2021-02-19T05:31:46.6264971Z         data_home : str or None
2021-02-19T05:31:46.6265305Z             Location to cache the response. None if no cache is required.
2021-02-19T05:31:46.6265610Z     
2021-02-19T05:31:46.6265833Z         Returns
2021-02-19T05:31:46.6266169Z         -------
2021-02-19T05:31:46.6266431Z         json_data : json
2021-02-19T05:31:46.6266780Z             the json result from the OpenML server if the call was successful.
2021-02-19T05:31:46.6267278Z             An exception otherwise.
2021-02-19T05:31:46.6267543Z         """
2021-02-19T05:31:46.6267744Z     
2021-02-19T05:31:46.6268006Z         @_retry_with_clean_cache(url, data_home)
2021-02-19T05:31:46.6268314Z         def _load_json():
2021-02-19T05:31:46.6268670Z             with closing(_open_openml_url(url, data_home)) as response:
2021-02-19T05:31:46.6269257Z                 return json.loads(response.read().decode("utf-8"))
2021-02-19T05:31:46.6269565Z     
2021-02-19T05:31:46.6269787Z         try:
2021-02-19T05:31:46.6270033Z             return _load_json()
2021-02-19T05:31:46.6270334Z         except HTTPError as error:
2021-02-19T05:31:46.6270703Z             # 412 is an OpenML specific error code, indicating a generic error
2021-02-19T05:31:46.6271046Z             # (e.g., data not found)
2021-02-19T05:31:46.6271353Z             if error.code != 412:
2021-02-19T05:31:46.6271653Z                 raise error
2021-02-19T05:31:46.6271892Z     
2021-02-19T05:31:46.6272156Z         # 412 error, not in except for nicer traceback
2021-02-19T05:31:46.6272482Z >       raise OpenMLError(error_message)
2021-02-19T05:31:46.6272866Z E       sklearn.datasets._openml.OpenMLError: Dataset with data_id 1461 not found.
2021-02-19T05:31:46.6273089Z 
2021-02-19T05:31:46.6273646Z /opt/hostedtoolcache/Python/3.9.1/x64/lib/python3.9/site-packages/sklearn/datasets/_openml.py:179: OpenMLError
```
</details>

**Comment by adrinjalali** [comment]
This is fixed by re-running the tests. Not ideal, but not a big issue. Hmm

**Comment by hildeweerts** [comment]
To avoid re-running all tests we can also hardcode a few re-tries to OpenML calls, perhaps with a short time delay. That usually fixed my problems with API calls during a benchmark project I did a few years ago.

**Comment by riedgar-ms** [comment]
This is causing issues with some of our internal testing as well. We are investigating whether we can host the Census dataset on a CDN (there is retry logic there, but we've had problems which have defeated that).

**Comment by romanlutz** [comment]
Definitely seems like an openml issue to me, and not because of removed datasets but unavailable service.
PapersWithCode also recently added datasets: https://paperswithcode.com/datasets

**Comment by romanlutz** [comment]
Thanks for linking @hildeweerts . Sounds like this should be resolved now. Can we close this issue?

**Comment by hildeweerts** [comment]
I think so. We can always re-open if we still encounter any issues.

**Comment by hildeweerts** [comment]
Reopened this issue (see https://github.com/fairlearn/fairlearn/pull/732#issuecomment-824583523) as well as the one the at OpenML repo (https://github.com/openml/OpenML/issues/1092).

@riedgar-ms you mentioned you've previously had some retry logic there, what was the reason that didn't work?

**Comment by riedgar-ms** [comment]
Most likely, the retry logic was only in the AzureML Notebooks - there was never a significant problem here. Can you point me at a build where this has been happening? I've just done a quick swing past, and I'm seeing mainly green.

**Comment by hildeweerts** [comment]
I don't know of a specific build, but @romanlutz mentioned you've encountered this issue several times over the past few weeks (see https://github.com/fairlearn/fairlearn/pull/732#issuecomment-824583523).

**Comment by romanlutz** [comment]
FWIW I haven't seen this in quite a while. Thanks @hildeweerts for raising it with OpenML folks, and obviously thanks to the OpenML folks who evidently made a few improvements on their side. 

@hildeweerts do you want to keep this open?

**Comment by hildeweerts** [comment]
@romanlutz let's close for now, if the issue pops up again we can always reopen :)

**Comment by romanlutz** [comment]
Reopening since I'm seeing tons of OpenML failures in our pipelines 🙁

**Comment by joaquinvanschoren** [comment]
Hi all, sorry about the hickups, we're doing a bunch of work on the backend migrating to Kubernetes, and occasionally response may be slower. We're definitely planning to resolve it completely.

**Comment by hildeweerts** [comment]
Thank you for letting us know, @joaquinvanschoren!

**Comment by TamaraAtanasoska** [comment]
We still have OpenML failures, is there any action that we can take? One thing suggested by @adrinjalali was to change the cascade close on jobs if one fails(mostly because of this reason), so just individual ones can be retriggered. I'd be also happy to look further into adding additional retry logic to CI if there is anything that can be added to help.

**Comment by riedgar-ms** [comment]
Not quite sure what you mean; sorry I've not been tracking Fairlearn very closely of late. Is the issue with OpenML refusing connections, or datasets going missing.

**Comment by hildeweerts** [comment]
As I don't see this getting resolved from the OpenML side anytime soon, I like @adrinjalali's suggestion and exploring additional retry logic would be amazing as well @TamaraAtanasoska!!

**Comment by TamaraAtanasoska** [comment]
> Not quite sure what you mean; sorry I've not been tracking Fairlearn very closely of late. Is the issue with OpenML refusing connections, or datasets going missing.

Usually CI fails because the tests/examples fail when they tray to fetch datasets from the website.

**Comment by joaquinvanschoren** [comment]
The server is indeed slow at the moment. It was running fine bit since last week it's been slow. Looking into it. 

Using an exponential backoff usually works really well, though. I believe that's also what sklearn does.

Also, if this is an option for you, you can now switch from downloading ARFF files to downloading Parquet files, which is much faster.

**Comment by TamaraAtanasoska** [comment]
I just had a chat with some of the scikit-learn maintainers. We are using `sklearn.datasets.fetch_openml` to retrieve the datasets, which has a backoff already implemented (although not exponential). It seems that additional to the caching implemented in `sklearn.datasets.fetch_openml` there is a version of caching involved that they use on the CI side that enables scikit-learn to [fetch once and share between jobs ](https://github.com/scikit-learn/scikit-learn/blob/main/.circleci/config.yml#L34-L49). I think trying that out is a good idea on our side too, I can prepare a PR to try it out and see if it improves the retrieval rate. We can try other things if it doesn't bring significant improvement. Does that sound good?

**Comment by riedgar-ms** [comment]
Sounds reasonable to me

**Comment by TamaraAtanasoska** [comment]
> I just had a chat with some of the scikit-learn maintainers. We are using `sklearn.datasets.fetch_openml` to retrieve the datasets, which has a backoff already implemented (although not exponential). It seems that additional to the caching implemented in `sklearn.datasets.fetch_openml` there is a version of caching involved that they use on the CI side that enables scikit-learn to [fetch once and share between jobs ](https://github.com/scikit-learn/scikit-learn/blob/main/.circleci/config.yml#L34-L49). I think trying that out is a good idea on our side too, I can prepare a PR to try it out and see if it improves the retrieval rate. We can try other things if it doesn't bring significant improvement. Does that sound good?

The idea suggested here has been merged with #1392, but to see if this issue can be closed we need to see to what extent does it help with the CI experience. I can observe and comment back after merging in a few PRs time :)

**Comment by joaquinvanschoren** [comment]
In the meantime we’ve also resolved the issue with the slow OpenML server
response. We were getting hammered by the Claude crawler :). The server is
now responding very fast again.


Joaquin Vanschoren, Ph.D., ir.
Associate Professor
Department of Mathematics and Computer Science
Eindhoven University of Technology (TU/e)
https://joaquinvanschoren.github.io/
